src/system_identity/generator.rs

```diff
@@ -0,0 +1,62 @@
+use thiserror::Error;
+
+use crate::{key::creator::Creator as KeyCreator, TokenRetriever};
+
+use super::iam_client::{IAMClient, SystemIdentityCreationResponseData};
+
+#[derive(Debug, Clone, Error)]
+pub enum SystemIdentityCreationError {
+    // This should probably leverage associated types for the involved traits, so we
+    // can gain automatic conversions via #[from] annotations...
+    #[error("error retrieving token: `{0}`")]
+    TokenRetriever(String),
+    #[error("error creating key pair: `{0}`")]
+    KeyPairCreator(String),
+    #[error("error creating system identity: `{0}`")]
+    IAMClient(String),
+}
+
+pub struct SystemIdentityGenerator<K, T, I>
+where
+    K: KeyCreator,
+    T: TokenRetriever,
+    I: IAMClient,
+{
+    pub(super) key_creator: K,
+    pub(super) token_retriever: T,
+    pub(super) iam_client: I,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct SystemIdentity {
+    pub name: String,
+    pub client_id: String,
+    pub pub_key: Vec<u8>,
+}
```

Comment:
Nit: Would it make sense to move this to the parent module? `system_identity::SystemIdentity` instead of `system_identity::generator::SystemIdentity`

Reply:
Done in https://github.com/newrelic/newrelic-auth-rs/pull/25/commits/268920dcc76f6f15d7f9518a1ce1a253078118e6!